modules/CbChargeSOMDriver/chargesom/CbChargeSOM.cpp

```diff
@@ -0,0 +1,260 @@
+        case types::cb_board_support::CPState::D:
+        case types::cb_board_support::CPState::E:
+            return types::cb_board_support::CPState::E;
+        default:
+            return types::cb_board_support::CPState::PilotFault;
+        }
+    }
+
+    if (voltage >= -10000 /* mV */) /* -10 V <= x < -2 V */
+        return types::cb_board_support::CPState::PilotFault;
+
+    if (voltage >= -11000 /* mV */) { /* -11 V <= x < -10 V (not standard) */
+        return previous_state == types::cb_board_support::CPState::F ? types::cb_board_support::CPState::F
+                                                                     : types::cb_board_support::CPState::PilotFault;
+    }
+
+    if (voltage >= -13000 /* mV */) /* -13 V <= x < -11 V */
+        return types::cb_board_support::CPState::F;
+
+    return types::cb_board_support::CPState::PilotFault; /* < -13 V */
+}
+
+double CbChargeSOM::cp_get_duty_cycle() const {
+    // this method must be called with `sc_mutex`already hold
+
+    return this->sc.duty_cycle;
+}
+
+void CbChargeSOM::cp_set_duty_cycle(double duty_cycle) {
+    std::scoped_lock lock(this->sc_mutex);
+
+    this->sc.duty_cycle = floorl(duty_cycle);
+
+    this->sync_with_hw();
+}
+
+bool CbChargeSOM::cp_is_nominal_duty_cycle() const {
+    // this method must be called with `sc_mutex`already hold
+
+    return 0 < this->sc.duty_cycle && this->sc.duty_cycle < 100;
+}
+
+bool CbChargeSOM::cp_is_enabled() {
+    std::scoped_lock lock(this->sc_mutex);
+
+    // FIXME
+    return true;
+}
+
+void CbChargeSOM::cp_disable() {
+    std::scoped_lock lock(this->sc_mutex);
+
+    // FIXME it is unclear atm if the current safety controller fw can disable CP
+    // so let's signal 100% for now
+    this->sc.duty_cycle = 100;
+
+    this->sync_with_hw();
+}
+
+types::board_support_common::Ampacity CbChargeSOM::get_ampacity(int& voltage) {
+    //std::scoped_lock lock(this->sc_mutex);
+
+    // FIXME
+    voltage = this->sc.pp_value * 3300 / 4096;
+
+    // map the measured value
+
+    if (voltage >= 3100)
+        return types::board_support_common::Ampacity::None; // no cable connected
+
+    if (voltage >= 2500)
+        return types::board_support_common::Ampacity::A_13;
+
+    if (voltage >= 2000)
+        return types::board_support_common::Ampacity::A_20;
+
+    if (voltage >= 1200)
+        return types::board_support_common::Ampacity::A_32;
+
+    if (voltage >= 700)
+        return types::board_support_common::Ampacity::A_63_3ph_70_1ph;
+
+    throw std::underflow_error(
+        "The measured voltage for the Proximity Pilot is out-of-range (U_PP: " + std::to_string(voltage) + " mV).");
+}
+
+void CbChargeSOM::set_allow_power_on(bool allow_power_on) {
+    std::scoped_lock lock(this->sc_mutex);
+
+    // repeat requesting the switch change until we succeed (infinitely for the moment)
+    do {
+        this->sc.hvsw1_hs = allow_power_on;
+
+        this->sync_with_hw();
+
```

Comment:
No sleep?

Reply:
Not at this point/layer. Will re-check whether a sleep in libcbuart is reasonable...